src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
 {{site.alert.end}}
 
 
+### Can't promote `this` {#this}
+
+**The cause:**
+You're trying to promote `this`,
+but type promotion for `this` is not yet supported.
+
+One common `this` promotion scenario is when writing [extension methods][].
+If the [`on` type][] of the extension method is a nullable type,
+you'd want to do a null check to see whether `this` is null:
+
+**Example:**
+
+{:.bad}
+```dart
+extension E on int? {
+  int? get i {
+    return this != null;
+  }
+}
+```
+
+**Message:**
+
+```nocode
+`this` can't be promoted. 
+```
+
+**Solution:**
+
+Create a local variable to hold the value of `this`, then perform the null check.
+
+{:.good}
+```dart
+extension E on int? {
+  int? get i {
+    final self = this;
+    return self != null;
+  }
+}
+```
+
+### Only private fields can be promoted {#private}
+
+**The cause:**
+You're trying to promote a field, but the field is not private.
+
+It’s possible for other libraries in your program
+to override public fields with a getter. Because
+[getters might not return a stable value](#not-field),
+and the compiler can't know what other libraries are doing,
+non-private fields cannot be promoted.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  final int? n; // CONTEXT
+  C(this.n);
+}
+
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // ERROR
+  }
+}
+```
+
+Here, the compiler produces an error message
+explaining that the `+` operator can't be called on `n`
+because it is potentially null.
+
+**Message:**
+
+```nocode
+'n' refers to a public field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Making the field private lets the compiler be sure that no outside libraries
+could possibly override its value, so it's safe to promote.
+
+{:.good}
+```dart
+class C {
+  final int? _n;
+  C(this.n);
```

Comment:
Since you've renamed the field, you need to update references to it. This line needs to be `C(this._n);`.

The `test` method below needs to be:
```dart
test(C c) {
  if (c._n != null) {
    print(c._n + 1); // OK
  }
}
```